ledger-api/grpc-definitions/com/daml/ledger/api/v1/events_service.proto

```diff
@@ -0,0 +1,364 @@
+  // This field never contains parties that are signatories.
+  // Required
+  repeated string observers = 9;
+
+  // The agreement text of the contract.
+  // We use StringValue to properly reflect optionality on the wire for backwards compatibility.
+  // This is necessary since the empty string is an acceptable (and in fact the default) agreement
+  // text, but also the default string in protobuf.
+  // This means a newer client works with an older sandbox seamlessly.
+  // Optional
+  google.protobuf.StringValue agreement_text = 6;
+
+  // The time as of which this contract is present on the ledger.
+  google.protobuf.Timestamp ledger_effective_time = 10;
+}
+
+
+// Records that a choice has been exercised on a target contract.
+// 
+// Changes compared to v1.ExercisedEvent
+//  - replaced child_event_ids with descendant_event_offset_upper_bound, which is stable under filtering
+//  - no `event_id`, as `EventOffset` structure communicates transaction scope and nesting
+//  - include `ledger_effective_time`, as that is contract metadata relevant for state evolution
+message ChoiceExercisedEvent {
+  reserved 1; // was 'event_id'
+
+  // The ID of the target contract.
+  // Must be a valid LedgerString (as described in ``value.proto``).
+  // Required
+  string contract_id = 2;
+
+  // The template of the target contract.
+  // Required
+  Identifier template_id = 3;
+
+  // The choice that's been exercised on the target contract.
+  // Must be a valid NameString (as described in ``value.proto``).
+  // Optional for ARCHIVALS_ONLY
+  string choice = 5;
+
+  // The argument the choice was made with.
+  // Optional for ARCHIVALS_ONLY
+  Value choice_argument = 6;
+
+  // The parties that made the choice.
+  // Each element must be a valid PartyIdString (as described in ``value.proto``).
+  // Optional for ARCHIVALS_ONLY
+  repeated string acting_parties = 7;
+
+  // If true, the target contract may no longer be exercised.
+  // Required
+  bool consuming = 8;
+
+  reserved 9; // removed field
+
+  // The parties that are notified of this event. The witnesses of an exercise
+  // node will depend on whether the exercise was consuming or not.
+  // If consuming, the witnesses are the union of the stakeholders and
+  // the actors.
+  // If not consuming, the witnesses are the union of the signatories and
+  // the actors. Note that the actors might not necessarily be observers
+  // and thus signatories. This is the case when the controllers of a
+  // choice are specified using "flexible controllers", using the
+  // ``choice ... controller`` syntax, and said controllers are not
+  // explicitly marked as observers.
+  // Each element must be a valid PartyIdString (as described in ``value.proto``).
+  // Optional for ARCHIVALS_ONLY
+  repeated string witness_parties = 10;
+
+  reserved 11; // was repeated string child_event_ids = 11;
+
+  // An upper bound that is larger or equal to the largest `LocalEventOffset` of any descendant event.
```

Comment:
For privacy reasons, I'd prefer to put "equal to" here. See the above comment on event offsets with Alice and Bob